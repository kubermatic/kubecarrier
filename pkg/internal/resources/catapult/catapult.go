/*
Copyright 2019 The KubeCarrier Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package catapult

import (
	"fmt"
	"strings"

	adminv1beta1 "k8s.io/api/admissionregistration/v1beta1"
	rbacv1 "k8s.io/api/rbac/v1"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
	"k8s.io/apimachinery/pkg/runtime/schema"
	"sigs.k8s.io/kustomize/v3/pkg/image"
	"sigs.k8s.io/kustomize/v3/pkg/types"
	"sigs.k8s.io/yaml"

	"github.com/kubermatic/kubecarrier/pkg/internal/kustomize"
	utilwebhook "github.com/kubermatic/kubecarrier/pkg/internal/util/webhook"
	"github.com/kubermatic/kubecarrier/pkg/internal/version"
)

// Config holds the config information to generate the kubecarrier controller manager setup.
type Config struct {
	// Name is the name of the Catapult instance.
	Name string
	// Namespace that the Catapult instance should be deployed into.
	Namespace string

	ManagementClusterKind, ManagementClusterVersion,
	ManagementClusterGroup, ManagementClusterPlural string
	ServiceClusterKind, ServiceClusterVersion,
	ServiceClusterGroup, ServiceClusterPlural string
	ServiceClusterName, ServiceClusterSecret string

	WebhookStrategy string
}

var k = kustomize.NewDefaultKustomize()

func Manifests(c Config) ([]unstructured.Unstructured, error) {
	v := version.Get()
	kc := k.ForHTTP(vfs)
	if err := kc.MkLayer("man", types.Kustomization{
		// "." needs to be replaced, because it's forbidden for Deployment and Pod names
		NamePrefix: strings.Replace(c.Name, ".", "-", -1) + "-",
		Namespace:  c.Namespace,
		Images: []image.Image{
			{
				Name:   "quay.io/kubecarrier/catapult",
				NewTag: v.Version,
			},
		},
		Resources: []string{"../default"},
		PatchesStrategicMerge: []types.PatchStrategicMerge{
			"manager_env_patch.yaml",
		},
	}); err != nil {
		return nil, fmt.Errorf("cannot mkdir: %w", err)
	}

	mutatingWebhookPath := utilwebhook.GenerateMutateWebhookPathFromGVK(schema.GroupVersionKind{
		Group:   c.ManagementClusterGroup,
		Version: c.ManagementClusterVersion,
		Kind:    c.ManagementClusterKind,
	})

	// Patch environment
	// Note:
	// we are not using *appsv1.Deployment here,
	// because some fields will be defaulted to empty and
	// interfere with the strategic merge patch of kustomize.
	managerEnv := map[string]interface{}{
		"apiVersion": "apps/v1",
		"kind":       "Deployment",
		"metadata": map[string]string{
			"name":      "manager",
			"namespace": "system",
		},
		"spec": map[string]interface{}{
			"template": map[string]interface{}{
				"spec": map[string]interface{}{
					"containers": []map[string]interface{}{
						{
							"name": "manager",
							"env": []map[string]interface{}{
								{
									"name":  "CATAPULT_MANAGEMENT_CLUSTER_KIND",
									"value": c.ManagementClusterKind,
								},
								{
									"name":  "CATAPULT_MANAGEMENT_CLUSTER_VERSION",
									"value": c.ManagementClusterVersion,
								},
								{
									"name":  "CATAPULT_MANAGEMENT_CLUSTER_GROUP",
									"value": c.ManagementClusterGroup,
								},
								{
									"name":  "CATAPULT_SERVICE_CLUSTER_KIND",
									"value": c.ServiceClusterKind,
								},
								{
									"name":  "CATAPULT_SERVICE_CLUSTER_VERSION",
									"value": c.ServiceClusterVersion,
								},
								{
									"name":  "CATAPULT_SERVICE_CLUSTER_GROUP",
									"value": c.ServiceClusterGroup,
								},
								{
									"name":  "CATAPULT_SERVICE_CLUSTER_NAME",
									"value": c.ServiceClusterName,
								},
								{
									"name":  "CATAPULT_SERVICE_CLUSTER_KUBECONFIG",
									"value": "/config/kubeconfig",
								},
								{
									"name":  "CATAPULT_MUTATING_WEBHOOK_PATH",
									"value": mutatingWebhookPath,
								},
								{
									"name":  "CATAPULT_WEBHOOK_STRATEGY",
									"value": c.WebhookStrategy,
								},
							},
							"volumeMounts": []map[string]interface{}{
								{
									"name":      "kubeconfig",
									"mountPath": "/config",
									"readOnly":  true,
								},
							},
						},
					},
					"volumes": []map[string]interface{}{
						{
							"name": "kubeconfig",
							"secret": map[string]interface{}{
								"secretName": c.ServiceClusterSecret,
							},
						},
					},
				},
			},
		},
	}
	managerEnvBytes, err := yaml.Marshal(managerEnv)
	if err != nil {
		return nil, fmt.Errorf("marshalling manager env patch: %w", err)
	}
	if err = kc.WriteFile("/man/manager_env_patch.yaml", managerEnvBytes); err != nil {
		return nil, fmt.Errorf("writing manager_env_patch.yaml: %w", err)
	}

	// Generate ClusterRole for component
	role := rbacv1.ClusterRole{
		TypeMeta: metav1.TypeMeta{
			APIVersion: "rbac.authorization.k8s.io/v1",
			Kind:       "ClusterRole",
		},
		ObjectMeta: metav1.ObjectMeta{
			Name: "manager",
		},
		Rules: []rbacv1.PolicyRule{
			{
				APIGroups: []string{c.ManagementClusterGroup},
				Resources: []string{c.ManagementClusterPlural},
				Verbs: []string{
					"create", "delete", "get", "list", "patch", "update", "watch"},
			},
			{
				APIGroups: []string{c.ManagementClusterGroup},
				Resources: []string{c.ManagementClusterPlural + "/status"},
				Verbs:     []string{"get", "patch", "update"},
			},
		},
	}
	roleBytes, err := yaml.Marshal(role)
	if err != nil {
		return nil, fmt.Errorf("marshalling cluster role: %w", err)
	}
	if err := kc.WriteFile("/rbac/cluster_role.yaml", roleBytes); err != nil {
		return nil, fmt.Errorf("writing /rbac/cluster_role.yaml: %w", err)
	}

	failurePolicyFail := adminv1beta1.Fail
	mutatingWebhookConfiguration := adminv1beta1.MutatingWebhookConfiguration{
		TypeMeta: metav1.TypeMeta{
			APIVersion: "admissionregistration.k8s.io/v1beta1",
			Kind:       "MutatingWebhookConfiguration",
		},
		ObjectMeta: metav1.ObjectMeta{
			Name: "mutating-webhook-configuration",
		},
		Webhooks: []adminv1beta1.MutatingWebhook{
			{
				Name: fmt.Sprintf("m%s.kubecarrier.io", strings.ToLower(c.ManagementClusterKind)),
				ClientConfig: adminv1beta1.WebhookClientConfig{
					CABundle: []byte("Cg=="),
					Service: &adminv1beta1.ServiceReference{
						Namespace: "system",
						Name:      "webhook-service",
						Path:      &mutatingWebhookPath,
					},
				},
				FailurePolicy: &failurePolicyFail,
				Rules: []adminv1beta1.RuleWithOperations{
					{
						Operations: []adminv1beta1.OperationType{
							adminv1beta1.Create,
							adminv1beta1.Update,
						},
						Rule: adminv1beta1.Rule{
							APIGroups:   []string{c.ManagementClusterGroup},
							APIVersions: []string{c.ManagementClusterVersion},
							Resources:   []string{c.ManagementClusterPlural},
						},
					},
				},
			},
		},
	}
	mutatingWebhookConfigurationBytes, err := yaml.Marshal(mutatingWebhookConfiguration)
	if err != nil {
		return nil, fmt.Errorf("marshalling MutatingWebhookConfiguration: %w", err)
	}
	if err := kc.WriteFile("/webhook/manifests.yaml", mutatingWebhookConfigurationBytes); err != nil {
		return nil, fmt.Errorf("writing /webhook/manifests.yaml: %w", err)
	}

	// execute kustomize
	objects, err := kc.Build("/man")
	if err != nil {
		return nil, fmt.Errorf("running kustomize build: %w", err)
	}
	return objects, nil
}
