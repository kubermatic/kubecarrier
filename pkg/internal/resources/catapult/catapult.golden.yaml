- apiVersion: rbac.authorization.k8s.io/v1
  kind: Role
  metadata:
    creationTimestamp: null
    labels:
      kubecarrier.io/role: catapult
    name: db-eu-west-1-catapult-manager
    namespace: test3000
  rules:
  - apiGroups:
    - kubecarrier.io
    resources:
    - serviceclusterassignments
    verbs:
    - get
    - list
    - watch
  - apiGroups:
    - kubecarrier.io
    resources:
    - serviceclusterassignments/status
    verbs:
    - get
- apiVersion: rbac.authorization.k8s.io/v1
  kind: ClusterRole
  metadata:
    creationTimestamp: null
    labels:
      kubecarrier.io/role: catapult
    name: db-eu-west-1-catapult-manager
  rules:
  - apiGroups:
    - eu-west-1.provider
    resources:
    - couchdbinternals
    verbs:
    - create
    - delete
    - get
    - list
    - patch
    - update
    - watch
  - apiGroups:
    - eu-west-1.provider
    resources:
    - couchdbinternals/status
    verbs:
    - get
    - patch
    - update
- apiVersion: v1
  kind: ServiceAccount
  metadata:
    labels:
      kubecarrier.io/role: catapult
    name: db-eu-west-1-catapult-sa
    namespace: test3000
- apiVersion: rbac.authorization.k8s.io/v1
  kind: ClusterRoleBinding
  metadata:
    labels:
      kubecarrier.io/role: catapult
    name: db-eu-west-1-catapult-manager
  roleRef:
    apiGroup: rbac.authorization.k8s.io
    kind: ClusterRole
    name: db-eu-west-1-catapult-manager
  subjects:
  - kind: ServiceAccount
    name: db-eu-west-1-catapult-sa
    namespace: test3000
- apiVersion: rbac.authorization.k8s.io/v1
  kind: RoleBinding
  metadata:
    labels:
      kubecarrier.io/role: catapult
    name: db-eu-west-1-catapult-manager
    namespace: test3000
  roleRef:
    apiGroup: rbac.authorization.k8s.io
    kind: Role
    name: db-eu-west-1-catapult-manager
  subjects:
  - kind: ServiceAccount
    name: db-eu-west-1-catapult-sa
    namespace: test3000
- apiVersion: rbac.authorization.k8s.io/v1
  kind: Role
  metadata:
    labels:
      kubecarrier.io/role: catapult
    name: db-eu-west-1-catapult-leader-election-role
    namespace: test3000
  rules:
  - apiGroups:
    - ""
    resources:
    - configmaps
    verbs:
    - get
    - list
    - watch
    - create
    - update
    - patch
    - delete
  - apiGroups:
    - ""
    resources:
    - configmaps/status
    verbs:
    - get
    - update
    - patch
  - apiGroups:
    - ""
    resources:
    - events
    verbs:
    - create
- apiVersion: rbac.authorization.k8s.io/v1
  kind: RoleBinding
  metadata:
    labels:
      kubecarrier.io/role: catapult
    name: db-eu-west-1-catapult-leader-election-rolebinding
    namespace: test3000
  roleRef:
    apiGroup: rbac.authorization.k8s.io
    kind: Role
    name: db-eu-west-1-catapult-leader-election-role
  subjects:
  - kind: ServiceAccount
    name: db-eu-west-1-catapult-sa
    namespace: test3000
- apiVersion: apps/v1
  kind: Deployment
  metadata:
    labels:
      control-plane: manager
      kubecarrier.io/role: catapult
    name: db-eu-west-1-catapult-manager
    namespace: test3000
  spec:
    replicas: 1
    selector:
      matchLabels:
        control-plane: manager
        kubecarrier.io/role: catapult
    template:
      metadata:
        labels:
          control-plane: manager
          kubecarrier.io/role: catapult
      spec:
        containers:
        - args:
          - --cert-dir=$(CERT_DIR)
          env:
          - name: CATAPULT_MANAGEMENT_CLUSTER_KIND
            value: CouchDBInternal
          - name: CATAPULT_MANAGEMENT_CLUSTER_VERSION
            value: v1alpha1
          - name: CATAPULT_MANAGEMENT_CLUSTER_GROUP
            value: eu-west-1.provider
          - name: CATAPULT_SERVICE_CLUSTER_KIND
            value: CouchDB
          - name: CATAPULT_SERVICE_CLUSTER_VERSION
            value: v1alpha1
          - name: CATAPULT_SERVICE_CLUSTER_GROUP
            value: couchdb.io
          - name: CATAPULT_SERVICE_CLUSTER_NAME
            value: eu-west-1
          - name: CATAPULT_SERVICE_CLUSTER_KUBECONFIG
            value: /config/kubeconfig
          - name: CATAPULT_MUTATING_WEBHOOK_PATH
            value: /mutate-eu-west-1-provider-v1alpha1-couchdbinternal
          - name: CATAPULT_WEBHOOK_STRATEGY
            value: ""
          - name: KUBERNETES_NAMESPACE
            valueFrom:
              fieldRef:
                fieldPath: metadata.namespace
          - name: CERT_DIR
            value: /tmp/k8s-webhook-server/serving-certs
          image: quay.io/kubecarrier/catapult:was not build properly
          livenessProbe:
            httpGet:
              path: /healthz
              port: readiness-port
          name: manager
          ports:
          - containerPort: 9443
            name: webhook-server
            protocol: TCP
          - containerPort: 9440
            name: readiness-port
            protocol: TCP
          readinessProbe:
            httpGet:
              path: /readyz
              port: readiness-port
          resources:
            limits:
              cpu: 100m
              memory: 30Mi
            requests:
              cpu: 100m
              memory: 20Mi
          volumeMounts:
          - mountPath: /config
            name: kubeconfig
            readOnly: true
          - mountPath: /tmp/k8s-webhook-server/serving-certs
            name: cert
            readOnly: true
        serviceAccountName: db-eu-west-1-catapult-sa
        terminationGracePeriodSeconds: 10
        volumes:
        - name: kubeconfig
          secret:
            secretName: eu-west-1-kubeconfig
        - name: cert
          secret:
            defaultMode: 420
            secretName: webhook-server-cert
- apiVersion: admissionregistration.k8s.io/v1beta1
  kind: MutatingWebhookConfiguration
  metadata:
    annotations:
      cert-manager.io/inject-ca-from: test3000/db-eu-west-1-catapult-serving-cert
    creationTimestamp: null
    labels:
      kubecarrier.io/role: catapult
    name: db-eu-west-1-catapult-mutating-webhook-configuration
  webhooks:
  - clientConfig:
      caBundle: Q2c9PQ==
      service:
        name: db-eu-west-1-catapult-webhook-service
        namespace: test3000
        path: /mutate-eu-west-1-provider-v1alpha1-couchdbinternal
    failurePolicy: Fail
    name: mcouchdbinternal.kubecarrier.io
    rules:
    - apiGroups:
      - eu-west-1.provider
      apiVersions:
      - v1alpha1
      operations:
      - CREATE
      - UPDATE
      resources:
      - couchdbinternals
- apiVersion: v1
  kind: Service
  metadata:
    labels:
      kubecarrier.io/role: catapult
    name: db-eu-west-1-catapult-webhook-service
    namespace: test3000
  spec:
    ports:
    - port: 443
      targetPort: 9443
    selector:
      control-plane: manager
      kubecarrier.io/role: catapult
- apiVersion: cert-manager.io/v1alpha2
  kind: Issuer
  metadata:
    labels:
      kubecarrier.io/role: catapult
    name: db-eu-west-1-catapult-selfsigned-issuer
    namespace: test3000
  spec:
    selfSigned: {}
- apiVersion: cert-manager.io/v1alpha2
  kind: Certificate
  metadata:
    labels:
      kubecarrier.io/role: catapult
    name: db-eu-west-1-catapult-serving-cert
    namespace: test3000
  spec:
    dnsNames:
    - db-eu-west-1-catapult-webhook-service.test3000.svc
    - db-eu-west-1-catapult-webhook-service.test3000.svc.cluster.local
    issuerRef:
      kind: Issuer
      name: db-eu-west-1-catapult-selfsigned-issuer
    secretName: webhook-server-cert
